controller/two_lane_queue.go

```diff
@@ -0,0 +1,163 @@
+// All the default methods operate on the fast queue, unless noted otherwise.
+type twoLaneQueue struct {
+	workqueue.RateLimitingInterface
+	slowLane workqueue.RateLimitingInterface
+	// consumerQueue is necessary to ensure that we're not reconciling
+	// the same object at the exact same time (e.g. if it had been enqueued
+	// in both fast and slow and is the only object there).
+	consumerQueue workqueue.Interface
+
+	name string
+
+	fastChan chan item
+	slowChan chan item
+}
+
+// Creates a new twoLaneQueue.
+func newTwoLaneWorkQueue(name string) *twoLaneQueue {
+	rl := workqueue.DefaultControllerRateLimiter()
+	tlq := &twoLaneQueue{
+		RateLimitingInterface: workqueue.NewNamedRateLimitingQueue(
+			rl,
+			name+"-fast",
+		),
+		slowLane: workqueue.NewNamedRateLimitingQueue(
+			rl,
+			name+"-slow",
+		),
+		consumerQueue: workqueue.NewNamed(name + "-consumer"),
+		name:          name,
+		fastChan:      make(chan item),
+		slowChan:      make(chan item),
+	}
+	tlq.runConsumer()
+	tlq.runProducers()
+	return tlq
+}
+
+func process(q workqueue.Interface, ch chan item) {
+	go func(q workqueue.Interface) {
+		for {
+			i, d := q.Get()
+			// If the queue is empty and we're shutting down — stop the loop.
+			if d {
+				break
+			}
+			ch <- item{i: i, sd: d}
+		}
+		// Sender closes the channel
+		close(ch)
+	}(q)
+}
+
+func (tlq *twoLaneQueue) runProducers() {
+	process(tlq.RateLimitingInterface, tlq.fastChan)
+	process(tlq.slowLane, tlq.slowChan)
+}
+
+func (tlq *twoLaneQueue) runConsumer() {
+	tlq.consumerQueue = workqueue.NewNamed(tlq.name + "consumer")
+	// Shutdown flags.
+	fast, slow := true, true
+	go func() {
+		for fast || slow {
+			// By default drain the fast lane.
+			if fast {
+				select {
+				case item, ok := <-tlq.fastChan:
+					if !ok {
+						fast = false
+						continue
+					}
+					tlq.consumerQueue.Add(item.i)
+					continue
+				default:
+				}
+			}
+			// Fast lane is empty, so wait for either channel to trigger. Since
+			// channels are picked randomly any of the lanes might trigger.
+			select {
+			case item, ok := <-tlq.fastChan:
+				if !ok {
+					fast = false
+					continue
+				}
+				tlq.consumerQueue.Add(item.i)
+			case item, ok := <-tlq.slowChan:
+				if !ok {
+					slow = false
+					continue
+				}
+				tlq.consumerQueue.Add(item.i)
+			}
+		}
+		// We know both queues are shutdown now. Stop the consumerQueue.
+		tlq.consumerQueue.ShutDown()
+	}()
+}
+
+// Shutdown implements workqueue.Interace.
+// Shutdown shuts down both queues.
+func (tlq *twoLaneQueue) ShutDown() {
+	tlq.RateLimitingInterface.ShutDown()
+	tlq.slowLane.ShutDown()
+}
+
+// Done implements workqueue.Interface.
+// Done marks the item as completed in all the queues.
+// NB: this will just re-enqueue the object on the queue that
+// didn't originate the object.
+func (tlq *twoLaneQueue) Done(i interface{}) {
+	tlq.consumerQueue.Done(i)
+	tlq.RateLimitingInterface.Done(i)
+	tlq.slowLane.Done(i)
```

Comment:
IMO this should just be:
```suggestion
	tlq.consumerQueue.Done(i)
```

Reply:
Done.